[MM-44063] Line chart for top channels

BenCookie95 opened the pull request:
#### Summary
Line graph for top channels

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-44063

#### Screenshots
![Screen Shot 2022-05-24 at 2 06 28 PM](https://user-images.githubusercontent.com/14115924/170103139-86c480e5-b30d-4b2e-81fb-395f14f4171b.png)

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
NONE
```

mkraft commented on components/activity_and_insights/insights/top_channels/top_channels_line_chart/top_channels_line_chart.tsx:
```diff
@@ -0,0 +1,205 @@
+    const getGraphData = useMemo(() => {
+        const data = sortGraphData;
+        const dataset = [];
+
```
What would you think about putting the colours in an array and then doing this in a loop?

```js
const colours = [theme.buttonBg, theme.onlineIndicator, theme.awayIndicator, theme.dndIndicator, theme.newMessageSeparator];

for (i = 0; i < 5; i++) {
        if (topChannels[i]) {
            const colour = colours[i];
            dataset.push({
                fillColor: colour,
                borderColor: colour,
                pointBackgroundColor: theme.buttonBg,
                pointBorderColor: colour,
                backgroundColor: 'transparent',
                pointRadius: 0,
                hoverBackgroundColor: colour,
                label: topChannels[i].display_name,
                data: data.values[topChannels[i].id],
                hitRadius: 10,
            });
        }
}
```

mkraft commented on components/activity_and_insights/insights/top_channels/top_channels_line_chart/top_channels_line_chart.tsx:
```diff
@@ -0,0 +1,205 @@
+                    callbacks: {
+                        label(tooltipItem, data) {
+                            const index = tooltipItem.datasetIndex;
+                            if (typeof index !== 'undefined' && data.datasets && data.datasets[index] && data.datasets[index].label) {
```
I think this is equivalent...
```suggestion
                            if (typeof index !== 'undefined' && data?.datasets[index]?.label) {
```

BenCookie95 commented:
@matthewbirtch & @furqanmlk The line chart is available here https://mattermost-server-pr-20131.test.mattermost.cloud/

matthewbirtch reviewed (changes requested):
Nice work @BenCookie95. It's coming along nicely. A few things I'm hoping we can tweak
1. Are we able style the tooltip to make a few adjustments?
    - update the font-family to use `Open Sans`
    - update the font-weight on the message count line to be '400' (normal) and change the opacity of the text color to 64% on this line
    - add 4px of space between the color square and the text label on the first line
    - Can we change the color square to a circle and remove the outline?
    - I think we might need to left-align everything since it looks quite off when multiple items are showing in the tooltip

<img width="189" alt="image" src="https://user-images.githubusercontent.com/2040554/172432064-f7ba01d3-5907-45c8-a6fa-fc0083dcbbbe.png">

2. Can we update the font used for the axis labels?
    - Should be using `Opens Sans` for the font-family
    - Can the font size for the labels be 10px?
    - Can we change the label color to use 72% opacity center channel color?

3. Are we able to modify the date format used on the x-axis? It's hard to read in this format YYYY-MM-DD. I'd like to see a friendlier format if possible:
    - If the date is within the current year, can we use Feb 15 (for example)
    - If the date is from a previous year, can we use Feb 15, 2021

4. Are we able to reduce the number of labels that show on the bottom axis any further? It can get a little hard to read at some screen sizes — particularly when there are a lot of hours shown for the day.
<img width="1008" alt="image" src="https://user-images.githubusercontent.com/2040554/172434859-7507fd3c-09c7-42b1-a6cb-98eabd533fec.png">

5. Can we make sure the time that displays follows the current user's timezone? It doesn't seem to be doing this right now since it's showing 16:00 as the most recent hour, but it's only 12:37 in my timezone.

6. Actually, we should also follow the user's preference for `Clock Display` (12-hour vs. 24-hour clock)

4. It seems that the tooltip gets the top cut off if there are a lot of items in it. I only see this when filtered on the `Last 28 days` 
<img width="880" alt="image" src="https://user-images.githubusercontent.com/2040554/172433481-49f209e3-5c2f-4628-b0ea-592cae454a21.png">

8. The Top Channels card seems to flicker a height change when switching between date filters. Not a big deal, but I wondered if there was any way to smooth that out.

BenCookie95 commented:
@matthewbirtch I've fixed everything I can, some comments on the remaining issues below:

1. Are we able style the tooltip to make a few adjustments?
    * Can we change the color square to a circle and remove the outline? - _That isn’t possible unless I build a custom tooltip. This could be something I do in the future though_

4. Are we able to reduce the number of labels that show on the bottom axis any further? It can get a little hard to read at some screen sizes — particularly when there are a lot of hours shown for the day. _For `today` I have updated the labels to be every 2nd value if there are 8-12 labels and every 4th value if there are more than 12. Is that ok?_
5. Can we make sure the time that displays follows the current user's timezone? It doesn't seem to be doing this right now since it's showing 16:00 as the most recent hour, but it's only 12:37 in my timezone. - _Martin is looking into this on the backend. _

matthewbirtch commented:
Thanks for resolving those items @BenCookie95. It's looking a lot better. A few more things I'm seeing now:
1. We should truncate the channel name in the tooltip. It makese the tooltip super-wide. 
<img width="616" alt="image" src="https://user-images.githubusercontent.com/2040554/172860836-cbefef22-a63c-46df-b7b1-bbb92b9a40ef.png">
2. This is unrelated to this PR, but we should create a follow-up to fix this: when I switch to other channels and come back to the Insights view, it's not remembering the filters that were previously applied (Team vs My Insights, and time period selection). We should keep that selection persistent. Right now it seems to reset to the default

BenCookie95 commented:
> Thanks for resolving those items @BenCookie95. It's looking a lot better. A few more things I'm seeing now:
> 
> 1. We should truncate the channel name in the tooltip. It makese the tooltip super-wide.
> 
> <img alt="image" width="616" src="https://user-images.githubusercontent.com/2040554/172860836-cbefef22-a63c-46df-b7b1-bbb92b9a40ef.png">
> 
> 2. This is unrelated to this PR, but we should create a follow-up to fix this: when I switch to other channels and come back to the Insights view, it's not remembering the filters that were previously applied (Team vs My Insights, and time period selection). We should keep that selection persistent. Right now it seems to reset to the default

I can fix the first point but number 2 would have to be addressed in a different ticket

BenCookie95 commented:
> Thanks for resolving those items @BenCookie95. It's looking a lot better. A few more things I'm seeing now:
> 
> 1. We should truncate the channel name in the tooltip. It makese the tooltip super-wide.
> 
> <img alt="image" width="616" src="https://user-images.githubusercontent.com/2040554/172860836-cbefef22-a63c-46df-b7b1-bbb92b9a40ef.png">
> 
> 2. This is unrelated to this PR, but we should create a follow-up to fix this: when I switch to other channels and come back to the Insights view, it's not remembering the filters that were previously applied (Team vs My Insights, and time period selection). We should keep that selection persistent. Right now it seems to reset to the default

@matthewbirtch I have truncated it at a max of 16 characters, let me know if you want that increased/decreased

matthewbirtch commented:
> I can fix the first point but number 2 would have to be addressed in a different ticket

Can you create a ticket for this so it doesn't get list @BenCookie95?

matthewbirtch reviewed (approved):
Haven't seen it on the test server, but I'm good to approve this knowing you've made the necessary changes

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/205976

furqanmlk reviewed (approved):
Line chart for top channels is working fine.
